ktor-server/ktor-server-core/jvm/src/io/ktor/util/Paths.kt

```diff
@@ -0,0 +1,98 @@
+package io.ktor.util
+
+/**
+ * Process path components such as `.` and `..`, replacing redundant path components including all leading.
+ * It also discards all reserved characters and component names that are reserved (such as `CON`, `NUL`).
+ */
+@InternalAPI
+fun List<String>.normalizePathComponents(): List<String> {
+    for (index in indices) {
+        val component = get(index)
+        if (component.shouldBeReplaced()) {
+            return filterComponentsImpl(index)
+        }
+    }
+
+    return this
+}
+
+private fun List<String>.filterComponentsImpl(startIndex: Int): List<String> {
+    val result = ArrayList<String>(size)
+    if (startIndex > 0) {
+        result.addAll(subList(0, startIndex))
+    }
+    result.addReplacedComponent(get(startIndex))
+    for (index in startIndex + 1 until size) {
+        val component = get(index)
+        if (component.shouldBeReplaced()) {
+            result.addReplacedComponent(component)
+        } else {
+            result.add(component)
+        }
+    }
+
+    return result
+}
+
+private fun MutableList<String>.addReplacedComponent(component: String) {
+    if (component.isEmpty() || component == "." || component == "~" || component.toUpperCase() in ReservedWords) return
+    if (component == "..") {
+        if (isNotEmpty()) {
+            removeAt(lastIndex)
```

Comment:
It's quite ambiguous to call remove in add method. Could we avoid that?

Reply:
Yes, rename to processComponent to avoid add prefix